{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from SEQlinkage.Core import *\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SEQLinkage\n",
    "\n",
    "> Collapsed Haplotype Pattern Method for Linkage Analysis of Next-Generation Sequencing Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Features\n",
    "- It can do linkage analysis on single variants and CHP markers.\n",
    "- It can deal with families from different population.\n",
    "- It can handle large-scale whole-genome linkage analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Pre-requisites\n",
    "\n",
    "Make sure you install the pre-requisited before running seqlink:\n",
    "\n",
    "\n",
    "```\n",
    "#install cstatgen\n",
    "conda install -c conda-forge xeus-cling\n",
    "conda install -c anaconda swig \n",
    "conda install -c conda-forge gsl\n",
    "pip install egglib\n",
    "git clone https://github.com/statgenetics/cstatgen.git\n",
    "cd cstatgen\n",
    "python setup.py install\n",
    "\n",
    "#install paramlink2\n",
    "R\n",
    "install.packages(\"paramlink2\")\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Install"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`pip install SEQLinkage`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: seqlink [-h] [--chp] --fam FILE --vcf FILE [--anno FILE] [--pop FILE]\n",
      "               [--included-vars FILE] [-b FILE] [-c P] [-o Name]\n",
      "               [--build STRING] [--freq INFO] [--chrom-prefix STRING]\n",
      "               [--run-linkage] [-K FLOAT] [--moi STRING] [-W FLOAT] [-M FLOAT]\n",
      "               [--theta-max FLOAT] [--theta-inc FLOAT]\n",
      "\n",
      "SEQLinkage V2, linkage analysis using sequence data\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Collapsed haplotype pattern method arguments:\n",
      "  --chp                 Generate CHP markers.\n",
      "  --fam FILE            Input pedigree and phenotype information in FAM\n",
      "                        format.\n",
      "  --vcf FILE            Input VCF file, bgzipped.\n",
      "  --anno FILE           Input annotation file from annovar.\n",
      "  --pop FILE            Input two columns file, first column is family ID,\n",
      "                        second column population information.\n",
      "  --included-vars FILE  Variants to be included for linkage analysis, if None,\n",
      "                        the analysis won't filter any variants. But you can\n",
      "                        still set AF cutoff by -c argment.\n",
      "  -b FILE, --blueprint FILE\n",
      "                        Blueprint file that defines regional marker (format:\n",
      "                        \"chr startpos endpos name avg.distance male.distance\n",
      "                        female.distance\").\n",
      "  -c P, --maf-cutoff P  MAF cutoff to define variants to be excluded from\n",
      "                        analyses. this is useful, if you need to analyse\n",
      "                        multiple population together.\n",
      "  -o Name, --output Name\n",
      "                        Output name prefix.\n",
      "  --build STRING        Reference genome version for VCF file.\n",
      "  --freq INFO           Info field name for allele frequency in VCF file.\n",
      "  --chrom-prefix STRING\n",
      "                        Prefix to chromosome name in VCF file if applicable,\n",
      "                        e.g. \"chr\".\n",
      "\n",
      "LINKAGE options:\n",
      "  --run-linkage         Perform Linkage analysis.\n",
      "  -K FLOAT, --prevalence FLOAT\n",
      "                        Disease prevalence. Default to 0.001.\n",
      "  --moi STRING          Mode of inheritance, AD/AR: autosomal\n",
      "                        dominant/recessive. Default to AD.\n",
      "  -W FLOAT, --wt-pen FLOAT\n",
      "                        Penetrance for wild type. Default to 0.01.\n",
      "  -M FLOAT, --mut-pen FLOAT\n",
      "                        Penetrance for mutation. Default to 0.9.\n",
      "  --theta-max FLOAT     Theta upper bound. Default to 0.5.\n",
      "  --theta-inc FLOAT     Theta increment. Default to 0.05.\n"
     ]
    }
   ],
   "source": [
    "!seqlink --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Linkage analysis on specific regions\n",
    "> Normally, the regions are gene regions. you can also use self-defined regions, such as promoter regions, enhancer regions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.run seqlink on CHP marker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;40;32mMESSAGE: Binary trait detected in [/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_ped.fam]\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: Namespace(chp=True, tfam='/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_ped.fam', vcf='/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_snps.vcf.gz', anno='testdata/test_chr1_anno.csv', pop='testdata/test_fam_pop.txt', included_vars=None, blueprint='testdata/test_blueprint_ext.txt', maf_cutoff=0.05, output='data/test_chp', build='hg38', freq='AF', chr_prefix=None, run_linkage=True, prevalence=0.001, inherit_mode='AD', wild_pen=0.01, muta_pen=0.9, theta_max=0.5, theta_inc=0.05)\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: 18 samples found in FAM file but not in VCF file:\u001b[0m\n",
      "\n",
      "\u001b[1;40;32mMESSAGE: 18 samples found in [/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_snps.vcf.gz]\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: Loading marker map from [testdata/test_blueprint_ext.txt] ...\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: 6 families with a total of 18 samples will be scanned for 12 pre-defined units\u001b[0m\n",
      "True True\n",
      "chr1:13080:G:C is not in annotation\n",
      "chr1:13115:G:T is not in annotation\n",
      "chr1:13080:G:C is not in annotation\n",
      "chr1:13115:G:T is not in annotation\n",
      "chr1:14662:T:C is not in annotation\n",
      "chr1:14861:G:A is not in annotation\n",
      "chr1:15837:T:C is not in annotation\n",
      "chr1:16074:G:A is not in annotation\n",
      "chr1:16225:C:G is not in annotation\n",
      "chr1:16360:C:T is not in annotation\n",
      "chr1:16734:TGGG:* is not in annotation\n",
      "chr1:17147:G:* is not in annotation\n",
      "chr1:17793:A:G is not in annotation\n",
      "chr1:17843:A:* is not in annotation\n",
      "SNVHap MIR6859-1@1,MIR6859-2@1,MIR6859-3@1,MIR6859-4@1\n",
      "\u001b[1;40;32mMESSAGE: write to pickle: data/test_chp/chr1result/chr1result0.pickle,Gene number:4,Time:0.00016164010390639304\u001b[0m\n",
      "create data/test_chp/chr1result/chr1result0_AFcutoff0.05_linkage.input\n",
      "create data/test_chp/chr1result/chr1result0_AFcutoff0.05_linkage.lods\n",
      "0.448735611513257\n",
      "create data/test_chp/chr1result/chr1result0_AFcutoff0.05_linkage.besthlod\n",
      "\u001b[1;40;32mMESSAGE: ============= Finish analysis ==============\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!seqlink --fam testdata/test_ped.fam --vcf testdata/test_snps.vcf.gz --anno testdata/test_chr1_anno.csv --pop testdata/test_fam_pop.txt --blueprint testdata/test_blueprint_ext.txt -c 0.05 -o data/test_chp --chp --run-linkage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.run seqlink on variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;40;32mMESSAGE: Binary trait detected in [/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_ped.fam]\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: Namespace(chp=False, tfam='/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_ped.fam', vcf='/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_snps.vcf.gz', anno='testdata/test_chr1_anno.csv', pop='testdata/test_fam_pop.txt', included_vars=None, blueprint='testdata/test_blueprint_ext.txt', maf_cutoff=0.05, output='data/test_var', build='hg38', freq='AF', chr_prefix=None, run_linkage=True, prevalence=0.001, inherit_mode='AD', wild_pen=0.01, muta_pen=0.9, theta_max=0.5, theta_inc=0.05)\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: 18 samples found in FAM file but not in VCF file:\u001b[0m\n",
      "\n",
      "\u001b[1;40;32mMESSAGE: 18 samples found in [/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_snps.vcf.gz]\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: Loading marker map from [testdata/test_blueprint_ext.txt] ...\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: 6 families with a total of 18 samples will be scanned for 12 pre-defined units\u001b[0m\n",
      "True False\n",
      "chr1:13080:G:C is not in annotation\n",
      "chr1:13115:G:T is not in annotation\n",
      "chr1:13080:G:C is not in annotation\n",
      "chr1:13115:G:T is not in annotation\n",
      "chr1:14662:T:C is not in annotation\n",
      "chr1:14861:G:A is not in annotation\n",
      "chr1:15837:T:C is not in annotation\n",
      "chr1:16074:G:A is not in annotation\n",
      "chr1:16225:C:G is not in annotation\n",
      "chr1:16360:C:T is not in annotation\n",
      "chr1:16734:TGGG:* is not in annotation\n",
      "chr1:17147:G:* is not in annotation\n",
      "chr1:17793:A:G is not in annotation\n",
      "chr1:17843:A:* is not in annotation\n",
      "\u001b[1;40;32mMESSAGE: write to pickle: data/test_var/chr1result/chr1result0.pickle,Gene number:4,Time:3.215630900942617e-05\u001b[0m\n",
      "create data/test_var/chr1result/chr1result0_AFcutoff0.05_linkage.input\n",
      "create data/test_var/chr1result/chr1result0_AFcutoff0.05_linkage.lods\n",
      "0.5998611375689507\n",
      "create data/test_var/chr1result/chr1result0_AFcutoff0.05_linkage.besthlod\n",
      "\u001b[1;40;32mMESSAGE: ============= Finish analysis ==============\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!seqlink --fam testdata/test_ped.fam --vcf testdata/test_snps.vcf.gz --anno testdata/test_chr1_anno.csv --pop testdata/test_fam_pop.txt --blueprint testdata/test_blueprint_ext.txt -c 0.05 -o data/test_var --run-linkage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### No annotation\n",
    "> If you don't have the annotation file. there is no need to add `--pop`. And `--freq` should be setted based on the `INFO` column in vcf file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;40;32mMESSAGE: Binary trait detected in [/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_ped.fam]\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: Namespace(chp=False, tfam='/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_ped.fam', vcf='/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_snps.vcf.gz', anno=None, pop=None, included_vars=None, blueprint='testdata/test_blueprint_ext.txt', maf_cutoff=0.05, output='data/test_var', build='hg38', freq='AF', chr_prefix=None, run_linkage=True, prevalence=0.001, inherit_mode='AD', wild_pen=0.01, muta_pen=0.9, theta_max=0.5, theta_inc=0.05)\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: 18 samples found in FAM file but not in VCF file:\u001b[0m\n",
      "\n",
      "\u001b[1;40;32mMESSAGE: 18 samples found in [/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_snps.vcf.gz]\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: Loading marker map from [testdata/test_blueprint_ext.txt] ...\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: 6 families with a total of 18 samples will be scanned for 12 pre-defined units\u001b[0m\n",
      "True False\n",
      "\u001b[1;40;32mMESSAGE: write to pickle: data/test_var/chrallresult/chrallresult0.pickle,Gene number:4,Time:8.800002332362863e-06\u001b[0m\n",
      "create data/test_var/chrallresult/chrallresult0_AFcutoff0.05_linkage.input\n",
      "create data/test_var/chrallresult/chrallresult0_AFcutoff0.05_linkage.lods\n",
      "0.5967844482511282\n",
      "create data/test_var/chrallresult/chrallresult0_AFcutoff0.05_linkage.besthlod\n",
      "\u001b[1;40;32mMESSAGE: ============= Finish analysis ==============\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!seqlink --fam testdata/test_ped.fam --vcf testdata/test_snps.vcf.gz --freq='AF' --blueprint testdata/test_blueprint_ext.txt -c 0.05 -o data/test_chp_na --chp --run-linkage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Whole-genome linkage analysis\n",
    "> if `--blueprint` is not provided, the genomic region will be seperated to pseudogenes with 1000 variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;40;32mMESSAGE: Binary trait detected in [/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_ped.fam]\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: Generate regions by annotation\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: Namespace(chp=True, tfam='/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_ped.fam', vcf='/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_snps.vcf.gz', anno='testdata/test_chr1_anno.csv', pop='testdata/test_fam_pop.txt', included_vars=None, blueprint=None, maf_cutoff=0.05, output='data/test_chp_wg', build='hg38', freq='AF', chr_prefix=None, run_linkage=True, prevalence=0.001, inherit_mode='AD', wild_pen=0.01, muta_pen=0.9, theta_max=0.5, theta_inc=0.05)\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: 18 samples found in FAM file but not in VCF file:\u001b[0m\n",
      "\n",
      "\u001b[1;40;32mMESSAGE: 18 samples found in [/mnt/vast/hpc/csg/yin/Github/linkage/SEQpy3/nbs/testdata/test_snps.vcf.gz]\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: separate chromosome to regions\u001b[0m\n",
      "\u001b[1;40;32mMESSAGE: 6 families with a total of 18 samples will be scanned for 1 pre-defined units\u001b[0m\n",
      "True True\n",
      "chr1:10146:ACC:* is not in annotation\n",
      "chr1:10206:ACCCTAACCCTAACCCTAACCCTAAC:A is not in annotation\n",
      "chr1:10231:CCCCTAACCCTAACCCTAAACCCTAAACCCTAACCCTAACCCTAACCCTAACCCTAACCCCAACCCCAACCCCAACCCCAACCCCAACCCCAACCCTAACCCCTAACCCTAACCCTAACCCTACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCCTAACCCCTAA:* is not in annotation\n",
      "chr1:10250:ACCCTAAAC:* is not in annotation\n",
      "chr1:10397:CCCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCCTAACCCTAACCCTAACCCTAACCCTCGCGGTACCCTCAG:* is not in annotation\n",
      "chr1:10403:A:ACCCTAACCCTAACCCTAACCCTAACCCTAAC is not in annotation\n",
      "chr1:10403:ACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAAC:* is not in annotation\n",
      "chr1:10421:ACCCTAACCCTAACCCTAAC:* is not in annotation\n",
      "chr1:10422:CCCTAACCCTAACCCTAA:* is not in annotation\n",
      "chr1:10427:ACCCTAACCCTAACCCCT:* is not in annotation\n",
      "chr1:10428:CCCTAACCCTAACCCCTAA:* is not in annotation\n",
      "chr1:10431:TAACCC:* is not in annotation\n",
      "chr1:10432:AACCCTAACC:* is not in annotation\n",
      "chr1:10433:ACCCTAACC:* is not in annotation\n",
      "chr1:10436:CTAACCCCT:* is not in annotation\n",
      "chr1:10437:TAACCCCTAACC:* is not in annotation\n",
      "chr1:10438:AAC:* is not in annotation\n",
      "chr1:10439:ACCCCT:* is not in annotation\n",
      "chr1:10440:C:CCCTAACCCAACCCTAACCCTAACCCTAACCCTAA is not in annotation\n",
      "chr1:10440:CCCCTAACCCTAACCCTAACCCTAACCCTCGCGGTACCCTCAG:* is not in annotation\n",
      "chr1:10442:CCT:* is not in annotation\n",
      "chr1:10443:CTAACCCTAACCCTAACCCT:* is not in annotation\n",
      "chr1:10444:TAACCC:* is not in annotation\n",
      "chr1:10446:ACCCT:* is not in annotation\n",
      "chr1:10449:CTAA:* is not in annotation\n",
      "chr1:10450:TAACCC:* is not in annotation\n",
      "chr1:10451:AACCCT:* is not in annotation\n",
      "chr1:10452:AC:* is not in annotation\n",
      "chr1:10453:CCCTAA:* is not in annotation\n",
      "chr1:10455:C:* is not in annotation\n",
      "chr1:10464:A:AC is not in annotation\n",
      "chr1:13080:G:C is not in annotation\n",
      "chr1:13115:G:T is not in annotation\n",
      "chr1:14662:T:C is not in annotation\n",
      "chr1:14861:G:A is not in annotation\n",
      "chr1:15837:T:C is not in annotation\n",
      "chr1:16074:G:A is not in annotation\n",
      "chr1:16225:C:G is not in annotation\n",
      "chr1:16360:C:T is not in annotation\n",
      "chr1:16734:TGGG:* is not in annotation\n",
      "chr1:17147:G:* is not in annotation\n",
      "chr1:17793:A:G is not in annotation\n",
      "chr1:17843:A:* is not in annotation\n",
      "chr1:30845:G:A is not in annotation\n",
      "chr1:30921:TCG:* is not in annotation\n",
      "chr1:30928:CTCATT:* is not in annotation\n",
      "chr1:31719:G:GA is not in annotation\n",
      "chr1:31719:GA:G is not in annotation\n",
      "chr1:40893:G:GA is not in annotation\n",
      "chr1:47840:CT:C is not in annotation\n",
      "chr1:48173:C:G is not in annotation\n",
      "chr1:49513:AAG:* is not in annotation\n",
      "chr1:49515:G:* is not in annotation\n",
      "chr1:51375:A:C is not in annotation\n",
      "chr1:52278:TTC:T is not in annotation\n",
      "chr1:54342:A:G is not in annotation\n",
      "chr1:54597:C:G is not in annotation\n",
      "chr1:54713:T:* is not in annotation\n",
      "chr1:54715:TCTTTC:* is not in annotation\n",
      "chr1:54716:C:* is not in annotation\n",
      "chr1:54717:TTTCTTTCTTTC:* is not in annotation\n",
      "chr1:54718:TTCTTTCTTTCTTTCTTTCTTTC:* is not in annotation\n",
      "chr1:54719:TC:* is not in annotation\n",
      "chr1:54722:T:* is not in annotation\n",
      "chr1:54723:TC:* is not in annotation\n",
      "chr1:54726:T:* is not in annotation\n",
      "chr1:54727:TC:* is not in annotation\n",
      "chr1:54727:TC:T is not in annotation\n",
      "chr1:54729:TTTC:* is not in annotation\n",
      "chr1:54729:TTTC:T is not in annotation\n",
      "chr1:54730:T:* is not in annotation\n",
      "chr1:54731:TC:* is not in annotation\n",
      "chr1:54732:CTT:* is not in annotation\n",
      "chr1:54735:TC:* is not in annotation\n",
      "chr1:54736:CTTTCTTTCTTTCTTTCTTTCTT:* is not in annotation\n",
      "chr1:54737:TTTCTTTCTTTCTTTCTTTC:* is not in annotation\n",
      "chr1:54749:TTTCTTTC:* is not in annotation\n",
      "chr1:54755:TCTTTCTTTCTTTCTTTCTTC:* is not in annotation\n",
      "chr1:54757:TTTCTTTCTTTCTTTCTTCC:* is not in annotation\n",
      "chr1:54763:TCTTTCTTTCTTC:* is not in annotation\n",
      "chr1:54768:CTTTCTTCCTCCT:* is not in annotation\n",
      "chr1:54768:CTTTCTTCCTCCT:C is not in annotation\n",
      "chr1:54773:TTCCTCC:* is not in annotation\n",
      "chr1:55294:C:T is not in annotation\n",
      "chr1:55414:T:A is not in annotation\n",
      "chr1:55569:T:A is not in annotation\n",
      "chr1:55920:G:A is not in annotation\n",
      "chr1:57118:G:C is not in annotation\n",
      "chr1:57356:T:C is not in annotation\n",
      "chr1:57877:A:G is not in annotation\n",
      "chr1:58493:C:T is not in annotation\n",
      "chr1:58703:AG:A is not in annotation\n",
      "chr1:58716:T:C is not in annotation\n",
      "chr1:59311:T:C is not in annotation\n",
      "chr1:59387:CA:C is not in annotation\n",
      "chr1:59622:CT:C is not in annotation\n",
      "chr1:60312:A:G is not in annotation\n",
      "chr1:60590:C:T is not in annotation\n",
      "chr1:61720:A:C is not in annotation\n",
      "chr1:62101:T:TCCAC is not in annotation\n",
      "chr1:62160:C:G is not in annotation\n",
      "chr1:62681:G:C is not in annotation\n",
      "chr1:62865:CTT:* is not in annotation\n",
      "chr1:63089:C:A is not in annotation\n",
      "chr1:63695:G:GA is not in annotation\n",
      "chr1:64619:G:A is not in annotation\n",
      "chr1:66214:ATATAATATATAT:* is not in annotation\n",
      "chr1:66217:TAA:* is not in annotation\n",
      "chr1:66218:AATATATATTATATAATAT:* is not in annotation\n",
      "chr1:66219:A:* is not in annotation\n",
      "chr1:66221:ATATATTATATAATATATATTATATTATATAATATATAATATAAATATAATATAAAT:* is not in annotation\n",
      "chr1:66222:TATATTATATAATA:* is not in annotation\n",
      "chr1:66223:ATATTATATAATATATATTATATTATATAATATATAATATAAATATAATATAAAT:* is not in annotation\n",
      "chr1:66224:TATTATATAATATATATTATATTATATAATATATAATATAAATATAATATAA:* is not in annotation\n",
      "chr1:66225:AT:* is not in annotation\n",
      "chr1:66226:TTATATAATATATATTATATTATATAATATATAATATA:* is not in annotation\n",
      "chr1:66227:TATATAATATATATTATATTATATAATATATAATATAA:* is not in annotation\n",
      "chr1:66231:TAATATATATTATATTATATAATATATAATATAAATATAATATAA:* is not in annotation\n",
      "chr1:66235:ATATATTATATTATATAATATATAATATAAATATAATATAAAT:* is not in annotation\n",
      "chr1:66236:TATATTATATTATATAATATATAATATAAATATAATATAA:* is not in annotation\n",
      "chr1:66238:TATTATATTATATAATATATAATATAAATATAATATAA:* is not in annotation\n",
      "chr1:66240:TTATATTATATAATATATAATATAAATATAATATAA:* is not in annotation\n",
      "chr1:66244:ATTATATAATATATAATATAAATATAATATAAATTATAT:* is not in annotation\n",
      "chr1:66250:TAATATATAATATAA:* is not in annotation\n",
      "chr1:66253:TATATAATATAAATATAATATAAATTATATTATATA:* is not in annotation\n",
      "chr1:66257:TAATATAAATATAATATAA:* is not in annotation\n",
      "chr1:66258:AATATAAATATAATATAAATTATATTATATAATATATAATATAAATATAATATAAATT:* is not in annotation\n",
      "chr1:66260:TATAAATATAATATAAATTATATTATATAATATATA:* is not in annotation\n",
      "chr1:66262:TAAATATAATATAA:* is not in annotation\n",
      "chr1:66263:A:* is not in annotation\n",
      "chr1:66264:A:* is not in annotation\n",
      "chr1:66273:TAA:* is not in annotation\n",
      "chr1:66275:AATTAT:* is not in annotation\n",
      "chr1:66278:TATATTATATAATATATAATATAAATATAATATAAATTATATAA:* is not in annotation\n",
      "chr1:66281:AT:* is not in annotation\n",
      "chr1:66282:T:* is not in annotation\n",
      "chr1:66283:TATATAATATATAATATAAATATA:* is not in annotation\n",
      "chr1:66288:AAT:* is not in annotation\n",
      "chr1:66295:A:* is not in annotation\n",
      "chr1:66299:TAA:* is not in annotation\n",
      "chr1:66305:TAATATAAATTATATAAATATAATATATATTTTATTATATAATATAATATATATTATATAAATATAATATATAAATTATATAATATAATATATATTATATAATATAATATATTTTATTATATA:* is not in annotation\n",
      "chr1:66308:TATAAATTATATAAATATA:* is not in annotation\n",
      "chr1:66310:TAA:* is not in annotation\n",
      "chr1:66312:AAT:* is not in annotation\n",
      "chr1:66331:ATATTT:* is not in annotation\n",
      "chr1:66333:ATTTTATT:* is not in annotation\n",
      "chr1:66349:TA:* is not in annotation\n",
      "chr1:66350:AATATATATTATATAAATATAATATATAAATTATATAATATAAT:* is not in annotation\n",
      "chr1:66352:TA:* is not in annotation\n",
      "chr1:66369:TA:* is not in annotation\n",
      "chr1:66370:AAT:* is not in annotation\n",
      "chr1:66374:TATAAATTATATAATATAATA:* is not in annotation\n",
      "chr1:66376:TAAA:* is not in annotation\n",
      "chr1:66377:A:* is not in annotation\n",
      "chr1:66378:AATTAT:* is not in annotation\n",
      "chr1:66385:TA:* is not in annotation\n",
      "chr1:66387:ATATAATATAT:* is not in annotation\n",
      "chr1:66390:TA:* is not in annotation\n",
      "chr1:66391:A:* is not in annotation\n",
      "chr1:66400:TATATA:* is not in annotation\n",
      "chr1:66404:TAATA:* is not in annotation\n",
      "chr1:66414:T:* is not in annotation\n",
      "chr1:66424:T:* is not in annotation\n",
      "chr1:66426:TAAATATATA:* is not in annotation\n",
      "chr1:66431:ATATAT:* is not in annotation\n",
      "chr1:66432:T:* is not in annotation\n",
      "chr1:66435:ATTATATTAT:* is not in annotation\n",
      "chr1:66447:AATATATATTTTATT:* is not in annotation\n",
      "chr1:66457:TTA:* is not in annotation\n",
      "chr1:66461:T:* is not in annotation\n",
      "chr1:66466:A:* is not in annotation\n",
      "chr1:66479:TATTTATAG:* is not in annotation\n",
      "chr1:66480:ATTT:* is not in annotation\n",
      "chr1:66481:TTTATAGA:* is not in annotation\n",
      "\u001b[1;40;32mMESSAGE: write to pickle: data/test_chp_wg/chr1result/chr1result0.pickle,Gene number:1,Time:0.00011989794775015778\u001b[0m\n",
      "create data/test_chp_wg/chr1result/chr1result0_AFcutoff0.05_linkage.input\n",
      "create data/test_chp_wg/chr1result/chr1result0_AFcutoff0.05_linkage.lods\n",
      "0.31306900456547737\n",
      "create data/test_chp_wg/chr1result/chr1result0_AFcutoff0.05_linkage.besthlod\n",
      "\u001b[1;40;32mMESSAGE: ============= Finish analysis ==============\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!seqlink --fam testdata/test_ped.fam --vcf testdata/test_snps.vcf.gz --anno testdata/test_chr1_anno.csv --pop testdata/test_fam_pop.txt -c 0.05 -o data/test_chp_wg --chp --run-linkage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- 1. `--fam`, Fam file (required, format: \"fid iid fathid mothid sex trait[1 control, 2 case, -9 or 0 missing]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ../testdata/test_ped.fam\n"
     ]
    }
   ],
   "source": [
    "%%writefile testdata/test_ped.fam\n",
    "1033    1033_2  0       0       2       -9\n",
    "1033    1033_1  0       0       1       -9\n",
    "1033    1033_99 1033_1  1033_2  2       1\n",
    "1033    1033_9  1033_1  1033_2  2       1\n",
    "1033    1033_3  1033_1  1033_2  2       2\n",
    "1036    1036_99 1036_1  1036_2  2       2\n",
    "1036    1036_6  0       0       1       2\n",
    "1036    1036_1  0       0       1       -9\n",
    "1036    1036_3  1036_6  1036_99 2       1\n",
    "1036    1036_4  1036_6  1036_99 2       1\n",
    "1036    1036_2  0       0       2       -9\n",
    "1036    1036_5  1036_6  1036_99 1       1\n",
    "10J_103 10J_103_10      0       0       1       -9\n",
    "10J_103 10J_103_4       0       0       1       -9\n",
    "10J_103 10J_103_3       0       0       2       -9\n",
    "10J_103 10J_103_2       10J_103_4       10J_103_3       2       2\n",
    "10J_103 10J_103_1       10J_103_10      10J_103_3       1       2\n",
    "10J_109 10J_109_2       10J_109_6       10J_109_5       1       2\n",
    "10J_109 10J_109_3       10J_109_6       10J_109_5       1       2\n",
    "10J_109 10J_109_4       10J_109_6       10J_109_5       1       2\n",
    "10J_109 10J_109_6       0       0       1       -9\n",
    "10J_109 10J_109_1       10J_109_6       10J_109_5       1       2\n",
    "10J_109 10J_109_5       0       0       2       2\n",
    "10J_109 10J_109_7       10J_109_6       10J_109_5       1       1\n",
    "10J_112 10J_112_3       0       0       1       1\n",
    "10J_112 10J_112_5       10J_112_3       10J_112_2       1       2\n",
    "10J_112 10J_112_1       10J_112_3       10J_112_2       2       1\n",
    "10J_112 10J_112_7       10J_112_3       10J_112_2       1       1\n",
    "10J_112 10J_112_2       0       0       2       2\n",
    "10J_119 10J_119_2       0       0       1       1\n",
    "10J_119 10J_119_5       0       0       2       1\n",
    "10J_119 10J_119_4       0       0       1       1\n",
    "10J_119 10J_119_6       10J_119_4       10J_119_5       1       2\n",
    "10J_119 10J_119_7       10J_119_4       10J_119_5       2       2\n",
    "10J_119 10J_119_1       10J_119_4       10J_119_5       2       2\n",
    "10J_119 10J_119_3       10J_119_2       10J_119_1       1       1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- 2. `--vcf`, VCF file (required, vcf.gz with vcf.gz.tbi)\n",
    "```\n",
    "bgzip -c file.vcf > file.vcf.gz\n",
    "tabix -p vcf file.vcf.gz\n",
    "```\n",
    "\n",
    "\n",
    "- 3. `--anno`, Annotation file from `ANNOVAR`, It must contains the allele frequency for population in the file of family population information. For example in here, The annotation file must have AF_amr, AF_afr, AF_nfe columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno=pd.read_csv('testdata/test_chr1_anno.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGene</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>GeneDetail.refGene</th>\n",
       "      <th>ExonicFunc.refGene</th>\n",
       "      <th>AAChange.refGene</th>\n",
       "      <th>...</th>\n",
       "      <th>CLNDISDB</th>\n",
       "      <th>CLNREVSTAT</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>DN ID</th>\n",
       "      <th>Patient ID</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Study</th>\n",
       "      <th>Pubmed ID</th>\n",
       "      <th>Otherinfo1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10140</td>\n",
       "      <td>10147</td>\n",
       "      <td>ACCCTAAC</td>\n",
       "      <td>A</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NONE;DDX11L1</td>\n",
       "      <td>dist=NONE;dist=1727</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:10140:ACCCTAAC:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10146</td>\n",
       "      <td>10147</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NONE;DDX11L1</td>\n",
       "      <td>dist=NONE;dist=1727</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:10146:AC:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10146</td>\n",
       "      <td>10148</td>\n",
       "      <td>ACC</td>\n",
       "      <td>*</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:10146:ACC:*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>10150</td>\n",
       "      <td>10151</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NONE;DDX11L1</td>\n",
       "      <td>dist=NONE;dist=1723</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:10150:CT:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10172</td>\n",
       "      <td>10177</td>\n",
       "      <td>CCCTAA</td>\n",
       "      <td>C</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NONE;DDX11L1</td>\n",
       "      <td>dist=NONE;dist=1697</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:10172:CCCTAA:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>1</td>\n",
       "      <td>66479</td>\n",
       "      <td>66487</td>\n",
       "      <td>TATTTATAG</td>\n",
       "      <td>*</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:66479:TATTTATAG:*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>1</td>\n",
       "      <td>66480</td>\n",
       "      <td>66481</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>FAM138A;OR4F5</td>\n",
       "      <td>dist=30399;dist=2610</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:66480:AT:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>1</td>\n",
       "      <td>66480</td>\n",
       "      <td>66483</td>\n",
       "      <td>ATTT</td>\n",
       "      <td>*</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:66480:ATTT:*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>1</td>\n",
       "      <td>66481</td>\n",
       "      <td>66488</td>\n",
       "      <td>TTTATAGA</td>\n",
       "      <td>T</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>FAM138A;OR4F5</td>\n",
       "      <td>dist=30400;dist=2603</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:66481:TTTATAGA:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>1</td>\n",
       "      <td>66481</td>\n",
       "      <td>66488</td>\n",
       "      <td>TTTATAGA</td>\n",
       "      <td>*</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:66481:TTTATAGA:*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 152 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chr  Start    End        Ref Alt Func.refGene   Gene.refGene  \\\n",
       "0      1  10140  10147   ACCCTAAC   A   intergenic   NONE;DDX11L1   \n",
       "1      1  10146  10147         AC   A   intergenic   NONE;DDX11L1   \n",
       "2      1  10146  10148        ACC   *            .              .   \n",
       "3      1  10150  10151         CT   C   intergenic   NONE;DDX11L1   \n",
       "4      1  10172  10177     CCCTAA   C   intergenic   NONE;DDX11L1   \n",
       "..   ...    ...    ...        ...  ..          ...            ...   \n",
       "995    1  66479  66487  TATTTATAG   *            .              .   \n",
       "996    1  66480  66481         AT   A   intergenic  FAM138A;OR4F5   \n",
       "997    1  66480  66483       ATTT   *            .              .   \n",
       "998    1  66481  66488   TTTATAGA   T   intergenic  FAM138A;OR4F5   \n",
       "999    1  66481  66488   TTTATAGA   *            .              .   \n",
       "\n",
       "       GeneDetail.refGene ExonicFunc.refGene AAChange.refGene  ... CLNDISDB  \\\n",
       "0     dist=NONE;dist=1727                  .                .  ...        .   \n",
       "1     dist=NONE;dist=1727                  .                .  ...        .   \n",
       "2                       .                  .                .  ...        .   \n",
       "3     dist=NONE;dist=1723                  .                .  ...        .   \n",
       "4     dist=NONE;dist=1697                  .                .  ...        .   \n",
       "..                    ...                ...              ...  ...      ...   \n",
       "995                     .                  .                .  ...        .   \n",
       "996  dist=30399;dist=2610                  .                .  ...        .   \n",
       "997                     .                  .                .  ...        .   \n",
       "998  dist=30400;dist=2603                  .                .  ...        .   \n",
       "999                     .                  .                .  ...        .   \n",
       "\n",
       "    CLNREVSTAT CLNSIG DN ID Patient ID Phenotype Platform Study Pubmed ID  \\\n",
       "0            .      .     .          .         .        .     .         .   \n",
       "1            .      .     .          .         .        .     .         .   \n",
       "2            .      .     .          .         .        .     .         .   \n",
       "3            .      .     .          .         .        .     .         .   \n",
       "4            .      .     .          .         .        .     .         .   \n",
       "..         ...    ...   ...        ...       ...      ...   ...       ...   \n",
       "995          .      .     .          .         .        .     .         .   \n",
       "996          .      .     .          .         .        .     .         .   \n",
       "997          .      .     .          .         .        .     .         .   \n",
       "998          .      .     .          .         .        .     .         .   \n",
       "999          .      .     .          .         .        .     .         .   \n",
       "\n",
       "                 Otherinfo1  \n",
       "0     chr1:10140:ACCCTAAC:A  \n",
       "1           chr1:10146:AC:A  \n",
       "2          chr1:10146:ACC:*  \n",
       "3           chr1:10150:CT:C  \n",
       "4       chr1:10172:CCCTAA:C  \n",
       "..                      ...  \n",
       "995  chr1:66479:TATTTATAG:*  \n",
       "996         chr1:66480:AT:A  \n",
       "997       chr1:66480:ATTT:*  \n",
       "998   chr1:66481:TTTATAGA:T  \n",
       "999   chr1:66481:TTTATAGA:*  \n",
       "\n",
       "[1000 rows x 152 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AF_amr</th>\n",
       "      <th>AF_afr</th>\n",
       "      <th>AF_nfe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0006</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0.0007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.6380</td>\n",
       "      <td>0.6300</td>\n",
       "      <td>0.6413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0357</td>\n",
       "      <td>0.0426</td>\n",
       "      <td>0.0370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0086</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>0.0084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>0.0039</td>\n",
       "      <td>0.0036</td>\n",
       "      <td>0.0076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>0.0163</td>\n",
       "      <td>0.0410</td>\n",
       "      <td>0.0279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     AF_amr  AF_afr  AF_nfe\n",
       "0    0.0006  0.0008  0.0007\n",
       "1    0.6380  0.6300  0.6413\n",
       "2         .       .       .\n",
       "3    0.0357  0.0426  0.0370\n",
       "4    0.0086  0.0097  0.0084\n",
       "..      ...     ...     ...\n",
       "995       .       .       .\n",
       "996  0.0039  0.0036  0.0076\n",
       "997       .       .       .\n",
       "998  0.0163  0.0410  0.0279\n",
       "999       .       .       .\n",
       "\n",
       "[1000 rows x 3 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anno.loc[:,['AF_amr', 'AF_afr', 'AF_nfe']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or, create a self-defined annotation file like this:\n",
    "```\n",
    "\tChr\tStart\tAF_amr\tAF\tAF_nfe\tAF_afr\n",
    "chr1:10140:ACCCTAAC:A\t1\t10140\t0.0006\t0.0007\t0.0007\t0.0008\n",
    "chr1:10146:AC:A\t1\t10146\t0.638\t0.6328\t0.6413\t0.63\n",
    "chr1:10150:CT:C\t1\t10150\t0.0357\t0.0375\t0.037\t0.0426\n",
    "chr1:10172:CCCTAA:C\t1\t10172\t0.0086\t0.0082\t0.0084\t0.0097\n",
    "chr1:10178:CCTAA:C\t1\t10178\t0.5\t0.3333\t0.2955\t0.4375\n",
    "chr1:10198:TAACCC:T\t1\t10198\t0.0\t0.0\t0.0\t0.0\n",
    "chr1:10231:C:A\t1\t10231\t0.2\t0.0366\t0.0\t0.05\n",
    "chr1:10236:AACCCT:A\t1\t10236\t0.0\t0.0\t0.0\t0.0\n",
    "chr1:10247:TAAACCCTA:T\t1\t10247\t0.2222\t0.2089\t0.1429\t0.4211\n",
    "```\n",
    "The index must match with the ID in vcf file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- 4. `--pop`, The file of family population information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing ../testdata/test_fam_pop.txt\n"
     ]
    }
   ],
   "source": [
    "%%writefile testdata/test_fam_pop.txt\n",
    "1033 AF_amr\n",
    "1036 AF_amr\n",
    "10J_103 AF_afr\n",
    "10J_109 AF_nfe\n",
    "10J_112 AF_nfe\n",
    "10J_119 AF_nfe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- 5. `--included-vars`, The file with one column of variants\n",
    "For example:\n",
    "```\n",
    "chr1:10140:ACCCTAAC:A\n",
    "chr1:10172:CCCTAA:C\n",
    "chr1:10198:TAACCC:T\n",
    "chr1:10236:AACCCT:A\n",
    "chr1:10261:T:TA\n",
    "chr1:10262:AACCCT:A\n",
    "```\n",
    "- 6. `--blueprint`, The blueprint file that defines regional marker (format: \"chr startpos endpos name avg.distance male.distance female.distance\"). The first four columns are required."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting testdata/test_blueprint_ext.txt\n"
     ]
    }
   ],
   "source": [
    "%%writefile testdata/test_blueprint_ext.txt\n",
    "1       11868   14362   LOC102725121@1  9.177127474362311e-07   1.1657192989882668e-06  6.814189157634088e-07\n",
    "1       11873   14409   DDX11L1 9.195320788455595e-07   1.1680302941673515e-06  6.82769803434766e-07\n",
    "1       14361   29370   WASH7P  1.5299877409602128e-06  1.94345806118021e-06    1.136044574393209e-06\n",
    "1       17368   17436   MIR6859-1@1,MIR6859-2@1,MIR6859-3@1,MIR6859-4@1 1.217692507120495e-06   1.5467668502473368e-06  9.041595098829462e-07\n",
    "1       30365   30503   MIR1302-10@1,MIR1302-11@1,MIR1302-2@1,MIR1302-9@1       2.1295973889038703e-06  2.705108741548526e-06   1.5812659765416382e-06\n",
    "1       34610   36081   FAM138A@1,FAM138C@1,FAM138F@1   2.4732411024120156e-06  3.1416201771056266e-06  1.8364278747737466e-06\n",
    "1       69090   70008   OR4F5   4.866641545668504e-06   6.181823219621424e-06   3.6135725636621673e-06\n",
    "1       134772  140566  LOC729737       9.633289838108921e-06   1.2236630588823159e-05  7.152898262617822e-06\n",
    "1       490755  495445  LOC100132062@1,LOC100132287@1   2.2828130832833112e-05  2.8997300893994373e-05  1.6950315013571593e-05\n",
    "1       450739  451678  OR4F16@1,OR4F29@1,OR4F3@1       2.575942360468604e-05   3.2720758549649544e-05  1.912685483821856e-05\n",
    "1       627379  629009  LOC101928626    3.943568768003252e-05   5.009295373297854e-05   2.9281737249900675e-05\n",
    "1       632614  632685  MIR12136        3.974742311959244e-05   5.048893386847169e-05   2.9513206656665908e-05"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting testdata/test_blueprint.txt\n"
     ]
    }
   ],
   "source": [
    "%%writefile testdata/test_blueprint.txt\n",
    "1       11868   14362   LOC102725121@1\n",
    "1       11873   14409   DDX11L1\n",
    "1       14361   29370   WASH7P\n",
    "1       17368   17436   MIR6859-1@1,MIR6859-2@1,MIR6859-3@1,MIR6859-4@1\n",
    "1       30365   30503   MIR1302-10@1,MIR1302-11@1,MIR1302-2@1,MIR1302-9@1\n",
    "1       34610   36081   FAM138A@1,FAM138C@1,FAM138F@1\n",
    "1       69090   70008   OR4F5\n",
    "1       134772  140566  LOC729737\n",
    "1       490755  495445  LOC100132062@1,LOC100132287@1\n",
    "1       450739  451678  OR4F16@1,OR4F29@1,OR4F3@1\n",
    "1       627379  629009  LOC101928626\n",
    "1       632614  632685  MIR12136"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
